### PR TITLE
QUICKSTEP-18: Allow BasicColumnStoreTupleStorageSubBlock to be unsorted

### PR DESCRIPTION
**Pull request by craig-chasseur**
This makes sorting an optional (instead of required) property of BasicColumnStoreTupleStorageSubBlock.

### CODE REVIEW

File: parser/CMakeLists.txt [hunk 1]
```diff
@@ -374,12 +374,15 @@ set_property(SOURCE ${CMAKE_CURRENT_BINARY_DIR}/SqlParser_gen.cpp APPEND PROPERT
...
+  set_property(TARGET quickstep_parser_SqlLexer APPEND_STRING PROPERTY COMPILE_FLAGS " -Wno-deprecated-register")
 endif()
 
 # GCC will make a warning for unsigned-signed comparisons which are inherent
 # in the lexer. For this, we turn off the sign compare.
-set_target_properties(quickstep_parser_SqlLexer PROPERTIES COMPILE_FLAGS "-Wno-sign-compare")
+CHECK_CXX_COMPILER_FLAG("-Wno-sign-compare" COMPILER_HAS_WNO_SIGN_COMPARE)
+if (COMPILER_HAS_WNO_SIGN_COMPARE)
+  set_property(TARGET quickstep_parser_SqlLexer APPEND_STRING PROPERTY COMPILE_FLAGS " -Wno-sign-compare")
+endif()
```
Comment (zuyu) [inline parser/CMakeLists.txt]: @craig-chasseur I think changes in this file are not related to this PR. What do you think of doing so in another separate PR? Thanks!

File: storage/BasicColumnStoreTupleStorageSubBlock.cpp [hunk 2]
```diff
@@ -183,26 +189,27 @@ bool BasicColumnStoreTupleStorageSubBlock::DescriptionIsValid(
...
-  }
-  const Type &sort_attribute_type = relation.getAttributeById(sort_attribute_id)->getType();
-  if (!ComparisonFactory::GetComparison(ComparisonID::kLess).canCompareTypes(sort_attribute_type,
-                                                                             sort_attribute_type)) {
-    return false;
+  // If a sort attribute is specified, check that it exists and can be ordered
+  // by LessComparison.
+  if (description.HasExtension(
+          BasicColumnStoreTupleStorageSubBlockDescription::sort_attribute_id)) {
+    attribute_id sort_attribute_id = description.GetExtension(
```
Comment (zuyu) [inline storage/BasicColumnStoreTupleStorageSubBlock.cpp]: Could we mark it `const`? Thanks!

File: storage/BasicColumnStoreTupleStorageSubBlock.cpp [hunk 3]
```diff
@@ -473,10 +483,10 @@ void BasicColumnStoreTupleStorageSubBlock::setAttributeValueInPlaceTyped(
 bool BasicColumnStoreTupleStorageSubBlock::deleteTuple(const tuple_id tuple) {
   DEBUG_ASSERT(hasTupleWithID(tuple));
 
-  if (!column_null_bitmaps_.elementIsNull(sort_column_id_)) {
-    if (column_null_bitmaps_[sort_column_id_].getBit(tuple)) {
-      --(header_->nulls_in_sort_column);
-    }
+  if (sort_specified_
+      && !column_null_bitmaps_.elementIsNull(sort_column_id_)
+      && column_null_bitmaps_[sort_column_id_].getBit(tuple)) {
```
Comment (zuyu) [inline storage/BasicColumnStoreTupleStorageSubBlock.cpp]: Code style:

Could we be consistent about where to wrap a line for a conditional expression? I think most of time we use the following style. Thanks!

```
  if (aaa &&
      bbb &&
      cc)
```

File: storage/BasicColumnStoreTupleStorageSubBlock.cpp [hunk 4]
```diff
@@ -575,6 +585,13 @@ predicate_cost_t BasicColumnStoreTupleStorageSubBlock::estimatePredicateEvaluati
 TupleIdSequence* BasicColumnStoreTupleStorageSubBlock::getMatchesForPredicate(
     const ComparisonPredicate &predicate,
     const TupleIdSequence *filter) const {
+  if (!sort_specified_) {
+    FATAL_ERROR(
```
Comment (zuyu) [inline storage/BasicColumnStoreTupleStorageSubBlock.cpp]: Could we use `DCHECK(sort_specified_)` from `glog` instead? Thanks!

File: storage/BasicColumnStoreTupleStorageSubBlock.cpp [hunk 5]
```diff
@@ -671,6 +688,8 @@ void BasicColumnStoreTupleStorageSubBlock::shiftNullBitmaps(
 // total size of tuples contained in this sub-block. It could be done with
 // less memory, although the implementation would be more complex.
 bool BasicColumnStoreTupleStorageSubBlock::rebuildInternal() {
+  DEBUG_ASSERT(sort_specified_);
```
Comment (zuyu) [inline storage/BasicColumnStoreTupleStorageSubBlock.cpp]: Please use `DCHECK` from `glog`. `DEBUG_ASSERT` has been depreciated.

**Comment by zuyu** [comment]
@craig-chasseur Please remove all the copyright updates, per discussion in the dev mail list below. Thanks!

> All of your copyright headers need to look exactly like this:
    http://www.apache.org/legal/src-headers.html#headers

> In fact, this will be one of the requirements on which your first
ASF release will be predicated. All the other Copyright statements
need to be coalesced and moved into the NOTICE file similar to:
    https://github.com/apache/incubator-hawq/blob/master/NOTICE

File: storage/BasicColumnStoreTupleStorageSubBlock.cpp (hunk 3 above)
Comment (craig-chasseur) [inline storage/BasicColumnStoreTupleStorageSubBlock.cpp]: That's not the style used in this file, nor in the rest of Quickstep generally. See lines 154 and 526 in the same file.

**Comment by craig-chasseur** [comment]
Completed requested changes.

**Comment by pateljm** [comment]
Thanks @craig-chasseur! LGTM. Ready to merge, but need you to squash all the commits into one. Can you rebase as per the instructions at: https://cwiki.apache.org/confluence/display/QUICKSTEP/Workflow+For+Committers

File: storage/BasicColumnStoreTupleStorageSubBlock.hpp [hunk 6]
```diff
@@ -211,11 +213,12 @@ class BasicColumnStoreTupleStorageSubBlock : public TupleStorageSubBlock {
...
 
   // Sort all columns according to ascending order of values in the sort
-  // column. Returns true if any reordering occured.
+  // column. Returns true if any reordering occured. This should only be called
+  // when 'sort_specified_' is true, otherwise the block is always "built".
   bool rebuildInternal();
 
   tuple_id max_tuples_;
-  bool sorted_;
+  bool sort_specified_, sorted_;
```
Comment (zuyu) [inline storage/BasicColumnStoreTupleStorageSubBlock.hpp]: Since we set `sort_specified_` in the constructor's initialization list, we could mark `const`.

**Comment by zuyu** [comment]
@craig-chasseur We are good to merge after addressing one minor issue above, squashing all commits into one, and rebasing on top of the current master branch. Thanks!

**Comment by zuyu** [comment]
@craig-chasseur Next time when squashing multiple commits, please also edit the commit messages so that the resultant commit message would be better than the combination of all commit messages. Thanks!

```
commit c1476d1e7637a943b00e104cdfd9498e3c7cfe03
Author: Craig Chasseur <spoonboy42@gmail.com>
Date:   Sun Jun 12 04:01:17 2016 -0700

    QUICKSTEP-18: Allow BasicColumnStoreTupleStorageSubBlock to be unsorted

    Review comments applied

    Revert build fix for Lexer (separate PR has been opened)
```
